### ISSUE TITLE
Abstraction for using Codecs and serialization for Kafka wire format

### ISSUE THREAD

**Issue by zmccoy** [compstak/KafkaStreams4s#45]
Our current codecs for Avro are purely for the Avro bytes format, but KafkaStreams4s should be able to read / write the wire format https://docs.confluent.io/current/schema-registry/serdes-develop/index.html#wire-format as to be able to work in a normal kafka system.  I haven't given a ton of thought to the abstractions in this case, but wanted to have this issue out there.

**Comment by zmccoy** [comment]
I'm poking around at this right now.

**Comment by LukaJCB** [comment]
Does any of this code help: https://github.com/fd4s/fs2-kafka/tree/master/modules/vulcan/src/main/scala/fs2/kafka/vulcan ?

**Comment by zmccoy** [comment]
@LukaJCB It does!  What isn't great is that that I think we'd need essentially that code's abstractions to make this usable.  At this moment I feel like I am working on a poor man's copy of that.  Each time I've tried to trim an inner class from that I understand why they did it heh.

**Comment by zmccoy** [comment]
I'm struggling with if we want to do pretty much exactly what they have so that you can create your schema registry safely etc with `KafkaStreams4s` or would we rely on someone using both libraries together and only using the Ser/De portions.  I can easily see applications only using `KafkaStreams4s` and not needing consumer/producer semantics outside of the streams api.

**Comment by LukaJCB** [comment]
maybe this could be moved to a module in vulcan itself? Not sure what the best coa is here, maybe we should create an issue in vulcan or fs2-kafka to see what they think

**Comment by zmccoy** [comment]
Yeah, I think that's the best choice to keep things without possible regression.  I can make an issue on fs2-kafka and tie it back here today.  I'm going to keep making a small example of what they're doing to see how it could tie into KafkaStreams4s along with working with both Avro implementations.  An issue I'm seeing so far is knowing whether a ser/de we are going to use is for a key or not which would change the inner kafka ser/de used.

**Comment by zmccoy** [comment]
Got a minimal version of what the fs2-kafka-encoding of this looks like going here: https://github.com/compstak/KafkaStreams4s/compare/master...zmccoy:kafka-avro-format?expand=1#diff-33e9d5e2212f25b9d29a6a7e756f130dR54-R55

I have some questions noted there as to how to tie the `serialize` and `deserialize` methods on the corresponding classes into how the streams portion works along with making sure the key serialization/deserialization is being used, hopefully implicitly, for the keys only.

**Comment by zmccoy** [comment]
The fix for this is going to need to wait for: https://github.com/compstak/KafkaStreams4s/pull/50 and then I should be able to open some stuff up :)

**Comment by keirlawson** [comment]
Hey, my team could really use this feature with out streams/vulcan integration, any way I could help out get this finished?

**Comment by zmccoy** [comment]
@keirlawson Hey!  I have it partially completed if you want to take that branch over, I got pulled off of doing this at $work and never got back around to it, which is totally my bad.  The current portion of it is here: https://github.com/compstak/KafkaStreams4s/compare/master...zmccoy:kafka-avro-format?expand=1.

**Comment by keirlawson** [comment]
Ok, will take a look, any pointers as to what is outstanding from your perspective?

**Comment by zmccoy** [comment]
Getting https://github.com/compstak/KafkaStreams4s/pull/50, in master now, merged into that branch which then allows for us to have our keys and values have different SerDes.  A lot of that should be following the types.  I'll need to dive back into it and see where I was at for a bit to have a better outline of what's really left.  The ergonomics of usage was after the portion of allowing keys and values separate SerDes and using them.

**Comment by keirlawson** [comment]
Apologies, I never got round to implementing this and our team have now moved away from Streams
